speculos/client.py

```diff
@@ -0,0 +1,181 @@
+from contextlib import contextmanager
+from PIL import Image, ImageChops
+from typing import Generator, Tuple
+import json
+import logging
+import requests
+import socket
+import subprocess
+import time
+
+logger = logging.getLogger("speculos-client")
+logger.setLevel(logging.INFO)
+
+
+def split_apdu(data: bytes) -> Tuple[bytes, int]:
+    assert len(data) >= 2
+    status = int.from_bytes(data[-2:], "big")
+    return data[:-2], status
+
+
+def screenshot_equal(path1: str, path2: str) -> bool:
+    """Compare two images and return True if they are equal."""
+
+    with Image.open(path1) as img1:
+        with Image.open(path2) as img2:
+            diff_img = ImageChops.difference(img1, img2)
+    return diff_img.getbbox() is None
+
+
+class ApduException(Exception):
+    def __init__(self, sw=0x6F00) -> None:
+        self.sw = sw
+
+    def __str__(self) -> str:
+        return f"Exception: invalid status 0x{self.sw:x}"
+
+
+class ApduResponse:
+    def __init__(self, response: requests.Response) -> None:
+        self.response = response
+
+    def receive(self) -> bytes:
+        assert self.response.status_code == 200
+        data, status = split_apdu(bytes.fromhex(self.response.json()["data"]))
+        if status != 0x9000:
+            raise ApduException(status)
+        return data
+
+
+class Api:
+    def __init__(self, api_url: str) -> None:
+        self.api_url = api_url
+        self.timeout = 2000
+        self.session = requests.session()
```

Comment:
`requests.session()` is deprecated (https://github.com/psf/requests/commit/1c13c5e6f35cf05c4de758714bc99919e9bbbb43) and `requests.Session()` should be used instead